graph/src/ipfs_client.rs

```diff
@@ -139,41 +145,42 @@ impl IpfsClient {
         }
     }
 
-    /// Calls stat for the given API route, and returns the total size of the object.
-    pub async fn stat_size(
-        &self,
-        api: StatApi,
-        mut cid: String,
-        timeout: Duration,
-    ) -> Result<u64, reqwest::Error> {
-        let route = format!("{}/stat", api.route());
-        if api == StatApi::Files {
-            // files/stat requires a leading `/ipfs/`.
-            cid = format!("/ipfs/{}", cid);
-        }
-        let url = self.url(&route, &cid);
-        let res = self.call(url, None, Some(timeout)).await?;
-        match api {
-            StatApi::Files => Ok(res.json::<FilesStatResponse>().await?.cumulative_size),
-            StatApi::Block => Ok(res.json::<BlockStatResponse>().await?.size),
-        }
+    /// To check the existence of a cid, we do a cat of a single byte.
+    pub async fn exists(&self, cid: &str, timeout: Option<Duration>) -> Result<(), IpfsError> {
+        self.call(self.cat_url("cat", cid, Some(1)), None, timeout)
+            .await?;
+        Ok(())
```

Comment:
I haven't thought through the code changes this requires, but wouldn't it be better to try and get the first block of a file, assuming optimistically that it exists, and only continuing with the client that gets that block first? They could maybe share an `AtomicBool` which gets set by the first one to get the block and the other clients then just stop doing work when the flag becomes `true`

Reply:
I'm not aware of anyone actually configuring multiple clients so this won't be called much, but getting the first block would be interesting, particularly if the file fits in a single block.